### PR TITLE
Control result rows demand, requesting one PartialResultSet by default

### LINKED ISSUE #9: Investigate changes needed in Client Library to support backpressure
Right now, client library requests one more partial result set whenever it processes the previous result set. 
This may well be the correct behavior, given that Spanner PartialResultSet does not correspond to a row. But perhaps this can be made more efficient by providing hints to GrpcResultSet for the number of rows to retrieve next.

### PR DESCRIPTION
**Pull request by elefeint**
`FlatMapIterable` keeps a buffer of a specified size. Because we cannot predict the relationship between a `PartialResultSet` and a spanner row, we'll default to requesting one `PartialResultSet` at a time.

Clients can modify the default batch size by setting a property, either on connection factory level (`partial_result_set_prefetch`) or on a specific Statement (`setPartialResultSetPrefetch`).

Fixes #9. Sort of.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 1]
```diff
@@ -105,11 +118,15 @@ public SpannerConnectionConfiguration build() throws IOException {
...
       }
 
-      return new SpannerConnectionConfiguration(
+      SpannerConnectionConfiguration configuration = new SpannerConnectionConfiguration(
           this.projectId,
           this.instanceName,
           this.databaseName,
           this.credentials);
+
+      configuration.partialResultSetPrefetch = this.partialResultSetPrefetch;
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: Should this param just be added to the constructor of SpannerConnectionConfiguration for consistency? Instead of directly setting the field; that way you can also keep `partialResultSetPrefetch` final.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java [hunk 2]
```diff
@@ -92,4 +87,23 @@ void setClient(Client client) {
     this.client = client;
   }
 
+  private SpannerConnectionConfiguration createConfiguration(ConnectionFactoryOptions options) {
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java]: This method can be made static.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 3]
```diff
@@ -32,6 +32,8 @@
 
   private final GoogleCredentials credentials;
 
+  private int partialResultSetPrefetch = 1;
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: `partialResultSetFetchSize`?

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 4]
```diff
@@ -77,6 +83,8 @@ public GoogleCredentials getCredentials() {
 
     private GoogleCredentials credentials;
 
+    private int partialResultSetPrefetch = 1;
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: I think it should be `Integer`. Otherwise, you're duplicating the default value in the builder and the actual configuration class where the field is optional.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 1 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: conditional on whether it was set

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java [hunk 5]
```diff
@@ -46,6 +46,9 @@
   /** Option name for GCP Spanner instance. */
   public static final Option<String> INSTANCE = Option.valueOf("instance");
 
+  public static final Option<Integer> PARTIAL_RESULT_SET_PREFETCH
+      = Option.valueOf("partial_result_set_prefetch");
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java]: maybe `partial_result_set_fetch_size`?

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java [hunk 6]
```diff
@@ -92,4 +87,23 @@ void setClient(Client client) {
...
+          .setDatabaseName(options.getRequiredValue(DATABASE))
+          .setCredentials(options.getValue(GOOGLE_CREDENTIALS));
+
+      if (options.hasOption(PARTIAL_RESULT_SET_PREFETCH)) {
+        configBuilder.setPartialResultSetPrefetch(
+            options.getValue(PARTIAL_RESULT_SET_PREFETCH));
+      }
+      return configBuilder.build();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java]: What's causing the `IOException`? Can we please add an exception message here?

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 7]
```diff
@@ -41,6 +41,8 @@
 
   private String sql;
 
+  private int partialResultSetPrefetch = 1;
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: Duplicating  this default yet again? In fact, this is probably the only place you need the default of `1` to exist.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 8]
```diff
@@ -98,6 +100,9 @@ public Statement bindNull(int i, Class<?> type) {
           if (signal.hasError()) {
             return Mono.error(signal.getThrowable());
           }
+          if (signal.isOnComplete()) {
+            return Mono.just(new SpannerResult(Flux.empty(), Mono.just(0)));
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: For this line, I saw in the javadoc for switchOnFirst that they have a caveat:
> It is advised to return a Publisher derived from the original Flux in all cases, as not doing so would keep the original Publisher open and hanging with a single request until the inner Publisher terminates or the whole Flux is cancelled.

https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#switchOnFirst-java.util.function.BiFunction-

Will returning a mono not based on the original flux be problematic? I think it should be okay if its already complete...? Otherwise, you could something like `flux.then(Mono.just(new SpannerResult(..)));`

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java (hunk 7 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: I was thinking it was messy to keep setting it, but then the alternative is creating a wrapper Integer object just to have the privilege of leaving it null...

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java [hunk 9]
```diff
@@ -100,7 +103,10 @@ public Batch createBatch() {
 
   @Override
   public Statement createStatement(String sql) {
-    return new SpannerStatement(this.client, this.session, this.currentTransaction, sql);
+    SpannerStatement statement
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: Should SpannerStatement also be made to simply accept the SpannerConnectionConfiguration object through constructor, rather than accept the prefetch through a setter? similar to how connection has access to the configuration.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java (hunk 6 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java]: Yeah. It's acquiring default credentials, since that involves the file system. Kind of annoying to have it here in the first place; that's why we discussed a provider at some point.

File: src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProviderTest.java [hunk 10]
```diff
@@ -95,6 +110,61 @@ public void getDriverReturnsSpanner() {
     assertThat(this.spannerConnectionFactoryProvider.getDriver()).isEqualTo(DRIVER_NAME);
   }
 
+  @Test
+  public void partialResultSetPrefetchPropagatesAsDemand() {
```
Comment (dzou) [inline src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProviderTest.java]: Nice test :+1:

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 3 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: That ambiguous "set"... Damn you, English, for being so concise!

Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: PartialResult SetFetchSize
PartialResultSet FetchSize
prsFetchSize is too cryptic...
Halp.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java (hunk 6 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProvider.java]: I'm going to move it closer to where the problem first occurs, and wrap it into a non-transient reosource exception.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 3 above)
Comment (meltsufin) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: setPartialResultSetSize and getPartialResultSetSize seem fine to me.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 1 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: Setters are convenient for optional properties; then the constructor can be more compact and only deal with required properties.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java (hunk 8 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: This is catching the case where the result flux was empty, so the very first signal received is on complete. Without this check, isDmlQuery fails.
But it's a good point, I'll use the chaining.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java (hunk 9 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: As discussed, let's keep the minimal configuration (optional setter), and create a special-purpose StatementConfiguration object if the number of options grows unwieldy.

File: src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProviderTest.java (hunk 10 above)
Comment (elefeint) [reply src/test/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactoryProviderTest.java]: It better be! I was stuck for two days writing it.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java [hunk 11]
```diff
@@ -102,8 +104,13 @@ public Batch createBatch() {
 
   @Override
   public Statement createStatement(String sql) {
-    return new SpannerStatement(this.client, this.session,
-        this.transactionContext, sql);
+    SpannerStatement statement
+        = new SpannerStatement(this.client, this.session, this.transactionContext, sql);
+
+    if (this.partialResultSetFetchSize != null) {
+      statement.setPartialResultSetPrefetch(this.partialResultSetFetchSize);
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: s/setPartialResultSetPrefetch/setPartialResultSetFetchSize for consistency

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java [hunk 12]
```diff
@@ -102,8 +104,13 @@ public Batch createBatch() {
 
   @Override
   public Statement createStatement(String sql) {
-    return new SpannerStatement(this.client, this.session,
-        this.transactionContext, sql);
+    SpannerStatement statement
+        = new SpannerStatement(this.client, this.session, this.transactionContext, sql);
+
+    if (this.partialResultSetFetchSize != null) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: If you made this field nullable, i.e. `Integer` everywhere, you wouldn't need this if statement.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 13]
```diff
@@ -97,19 +106,35 @@ public Builder setCredentials(GoogleCredentials credentials) {
...
       }
 
-      return new SpannerConnectionConfiguration(
+      SpannerConnectionConfiguration configuration = new SpannerConnectionConfiguration(
           this.projectId,
           this.instanceName,
           this.databaseName,
           this.credentials);
+
+      if (this.partialResultSetFetchSize != null) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: Why the `if` statement here?

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactory.java [hunk 14]
```diff
@@ -38,9 +38,15 @@ public SpannerConnectionFactory(Client client, SpannerConnectionConfiguration co
   }
 
   @Override
-  public Publisher<SpannerConnection> create() {
+  public Mono<SpannerConnection> create() {
     return this.client.createSession(this.config.getFullyQualifiedDatabaseName())
-      .map(session -> new SpannerConnection(this.client, session));
+      .map(session -> {
+        SpannerConnection connection = new SpannerConnection(this.client, session);
+        if (this.config.getPartialResultSetFetchSize() != null) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionFactory.java]: Again, I don't think the `if` is needed here.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java [hunk 15]
```diff
@@ -180,4 +186,13 @@ private boolean isSelectQuery() {
   private static boolean isDmlQuery(PartialResultSet firstPartialResultSet) {
     return firstPartialResultSet.getMetadata().getRowType().getFieldsList().isEmpty();
   }
+
+  /**
+   * Allows customizing the number of {@link PartialResultSet} objects to request at a time.
+   * @param prefetch batch size to request from Cloud Spanner
+   */
+  public void setPartialResultSetPrefetch(int prefetch) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: Need to rename this.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java (hunk 7 above)
Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: I think this is unavoidable because this setting is propagated from the connectionFactory options

ConnectionFactory --> Connection --> Statement

Thus both Connection and Statement have the setting as a field to pass it downwards.

And I would not recommend the `Integer` approach because of confusing side effects.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java (hunk 12 above)
Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: Can we just leave this `int`? Moving forward, I don't think we'll be remembering the reason it was `Integer` and it seems like an unusual practice.

Comment (dzou) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnection.java]: Edit: nevermind

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java (hunk 7 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java]: going to make it integer everywhere except the getter, and add tests to catch null autoboxing
